### DISCUSSION TITLE
Moving forward from Atom sunset

### DISCUSSION THREAD

**Discussion by ghost** [atom-ide-community/atom#361] [Ideas]
[Atom official sunset is 15th December, 2022](https://github.blog/2022-06-08-sunsetting-atom/). If people would like to keep using atom, there will need to be an active fork. I believe this is the best place for people interested in continuing the project to congregate and begin discussing a gameplan for reigniting the community. 

https://github.com/atom/atom/issues/25510#issuecomment-1151966415

Linking/Quoting comment by @DeeDeeG in Atom/Atom below. They present a good foundation for what this fork will need to survive.

___
<td class="d-block comment-body markdown-body  js-comment-body">
          <p dir="auto">I'd like to see an active healthy Atom fork. I believe that calls for a team that can handle it.</p>
<p dir="auto"><strong>Skills needed</strong> for a team wanting to successfully fork Atom:</p>
<ul dir="auto">
<li>General JavaScript skills (or a willingness to learn and some prior programming experience in another language)
<ul dir="auto">
<li>The Atom core repo is pretty much a 100% JavaScript codebase.</li>
<li>Some familiarity with the NodeJS and Electron ecosystems is a plus, can also be learned along the way, but it is a lot</li>
<li>Familiarity with web dev is a plus (JS, HTML, a bit of CSS) for the front-end parts of Atom and its packages, as well as to better understand the Chromium/"browser" aspects of Electron</li>
</ul>
</li>
<li>C++ skills (perhaps surprisingly?), for various native packages Atom depends on that need maintenance and updates</li>
<li>A willingness to learn about Atom's many default/bundled dependencies and packages (many lines of code across many repos, requiring some knowledge of many programming languages, such as the various syntax highlighting support packages for all the programming languages).
<ul dir="auto">
<li>These packages are a part of the Atom experience. And their tests may break in CI when changing Atom core. Not breaking existing packages is an important part of a successful Atom fork, IMO.</li>
</ul>
</li>
<li>Familiarity with CI and tests (gotta keep the app from breaking)</li>
<li>Open source maintainership
<ul dir="auto">
<li>Practical skills to manage repos and handle community contributions</li>
<li>Diplomacy and a willingness to work through the politics/personal opinions that inevitably rise up</li>
<li>Time, time time, effort effort effort, lots of both. Atom is a big project.</li>
</ul>
</li>
</ul>
<p dir="auto">(Note: <em>No one person needs to have all these skills!</em> Any one of the above is helpful. I personally believe a team with all of those skills collectively is pretty much a requirement, though.)</p>
<p dir="auto"><strong>Major development to-do items</strong>:</p>
<ul dir="auto">
<li>Get CI working again, particularly Windows CI on newer Windows CI images
<ul dir="auto">
<li>(Atom's Windows CI currently runs in an older Windows image, which is being deprecated/discontinued, but the tests fail in the newer Windows images at the moment)</li>
</ul>
</li>
<li>Continue the forward march of Electron updates</li>
<li>General maintenance
<ul dir="auto">
<li>Work on bug fixes for issues users report</li>
<li>Review PRs from the community (if there are any/if the fork would get really big and active)</li>
<li>Update dependencies, including the many default/bundled packages (this is a huge amount of code across a huge number of repos)</li>
</ul>
</li>
</ul>
<p dir="auto"><strong>Unclear if needed</strong>, but the fork may need to re-implement or replace the functionality of the atom.io website:</p>
<ul dir="auto">
<li>Study and re-implement (or code an alternative to) the functionality of <code class="notranslate">atom.io</code>, in case that site goes down.
<ul dir="auto">
<li><code class="notranslate">atom.io</code> hosts Atom's package ecosystem, including the API that <code class="notranslate">apm</code> interacts with to publish or download/install packages</li>
<li>atom.io is also involved in the automatic Atom app updates on macOS and Windows (I think)</li>
<li>And it is involved in some additional user flows, such as getting an auth token for the <code class="notranslate">github</code> package. (The bundled/default package that handles git and GitHub integration in Atom.)</li>
<li>Would be nice to have a proper website for the fork, for continuity with Atom's "accessible" and "welcoming" branding and project spirit.</li>
</ul>
</li>
</ul>
<p dir="auto">Sorry for the long post. I hope that if there is passion for a fork, that this can point people in the right direction.</p>
      </td>

**Comment by confused-Techie** [comment]
I have some JavaScript skills to put where needed if those could be identified. Additionally some server hosting resources.

Although I can't find Atom's repo for their website API, which may not be opensource. But as I understand it the main functionality for the website:

* Display standard WebPages and Docs
* API for:
  - Searching Packages via APM
  - Downloading Packages via APM
  - Staring Packages via APM
  - Uploading Packages via APM
* Then like mentioned here in some way relating to auth tokens for github.

But it seems to my understanding that the packages themselves are never hosted on the Atom.io website instead returning links to the Git repo, and Atom.io only stores a modified package.json file. Which you can see by visiting [https://atom.io/api/packages](https://atom.io/api/packages). 

If this is the case I would be happy to get started on a functioning website, and could likely be done quite quickly if making it in Golang. Then try to keep it compliant with the existing [API Calls](https://flight-manual.atom.io/atom-server-side-apis/sections/atom-package-server-api/).

But am unclear in what way it integrates with Github and Auth in general for any requests that must be authenticated.

If anyone could shed some light or confirm the existing maintainers are not working on this I could get started.

**Reply by ghost** [reply]
[Here's a manual page on the atom package server API](https://flight-manual.atom.io/atom-server-side-apis/sections/atom-package-server-api/)

It may be wise to begin the process of archiving all static content hosted on ```*.atom.io``` in the face of the sunsetting process.



I cannot find any source code for this API (and don't have much time to dig, admittedly).

I think it is a good idea to begin reverse engineering atom.io *sooner rather than later*, including the package server API functionality.

**Reply by confused-Techie** [reply]
I'll begin doing what I can to archive all of the content including existing Packages that are there. Once we have the data we can take as long as needed to figure out how to properly manipulate it. 

Really though I'll prioritize getting the packages and what little server API documentation there is. But hopefully until December is enough to grab all content.

If anybody familiar with the APM Repo is able to shed some light on this. According to the [Publishing Docs on Atom.io](https://flight-manual.atom.io/hacking-atom/sections/publishing/)

>Now let's review what the apm publish command does:
> * Registers the package name on atom.io if it is being published for the first time.
> * Updates the version field in the package.json file and commits it.
> * Creates a new [Git tag](https://git-scm.com/book/en/Git-Basics-Tagging) for the version being published.
> * Pushes the tag and current branch up to GitHub.
> * Updates atom.io with the new version being published.

I would be interested to find out what part of this Atom.io is actually responsible for. Since in the documentation it also says:
> In fact, Atom itself shells out to apm rather than hitting the API directly. 

And finally luckily it looks like Atom.io uses Github for all authentication which allows the API to not worry about it as much, only passing the authentication data along.

**Reply by DeeDeeG** [reply]
From apm's side, interaction between apm and the packages API when publishing a package (`apm publish` subcommand) is done in this file here: https://github.com/atom/apm/blob/master/src/publish.coffee

(Thankfully someone left a lot of comments in that file.) Likewise, other apm sub-comands (like `apm star` for example) each have their own files in that dir of the apm repo.

**Reply by DeeDeeG** [reply]
The GitHub package just benefits from this nifty page (https://github.atom.io/login) that pops out an OAuth token, which lets the GitHub package auth to the [GitHub.com API](https://docs.github.com/en/rest), so you can push/pull commits to repos, and browse/interact with PRs from within Atom.

(See https://github.com/atom/github/issues/2686 for some insights. This functionality can be manually worked around/replaced by end users even today. Workaround is a bit of a chore and thus not user-friendly, though.)

I had merely _speculated_ there might be other packages with their own more-than-informational page like that... (Maybe [teletype](https://teletype.atom.io/) has similar? ~~I never used Teletype, so I'm not sure if it needs any authentication, honestly.~~ It does. ([this page](https://teletype.atom.io/login) and [this](https://teletype.atom.io/auth/teletype/token).))

**Reply by confused-Techie** [reply]
@DeeDeeG fantastic information. I'll look at that as soon as I can.

Also I wrote a simple script to back up the packages on Atom.io if anybody that has some disk space and network bandwidth to spare, I would love to not be the only failsafe for this data. The script is downloadable via [NPM](https://www.npmjs.com/package/@confused-techie/atompackagesarchive), and hosted on [Github](https://github.com/confused-Techie/AtomPackagesArchive)

It currently just backs up the paginated data from the basic /api/packages grabbing every single page.

But if anybody has commits to grab other data, especially a good one to grab the standard webpages and all required scripts I would appreciate the pull request.

EDIT:
I have updated this program recently to expand on its capabilities, and fix a few present issues. Its now ready to go to backup all packages on Atom.io in different formats to help replicate its capability later on as best as possible.

**Reply by DeeDeeG** [reply]
@confused-Techie Thanks for taking the initiative on this. At minimum, archiving the data ensures no knowledge is lost if atom.io goes down.

I had some problems with execution stopping with errors while using your script earlier today. (I am guessing it would be due to "500 internal server error" responses that I got somewhat regularly from the API server?) Reading your edit, it sounds like this has been solved for your archiving tool.

In any case, I am a big fan of bash scripts, so I went ahead and did a bash script to work around the 500 internal server error situation.

<details><summary><b>Here is my bash script to back up the package metadata, with way too many comments. (Click to expand.)</b></summary>

```sh
#!/usr/bin/env bash

echo "Fetching Atom's package list!"

for page_num in {1..416}; # The second number for this range should be the number of pages existing in the API's database. As of my writing this script, there are 416 pages of results. Yes, the first page is 1. There is no page 0.
do

 attempts=1 # Setting up some vars for tracking the status of attempting to fetch this page...
 curl_exit_status=1

 mkdir -p ${page_num} # Make a new folder for just this page of package results

 while test $curl_exit_status -ne 0; # We're going to retry however many times until curl thinks we got a successful HTTP response code
 do
  echo # Echo a blank line here for nicer output formatting...
  echo "Fetching Page $page_num"
  echo "Number of attempts for this page is: ${attempts}"
  curl "https://atom.io/api/packages?sort=created_at&page=${page_num}" -o "./${page_num}/packages.json" -D "./${page_num}/headers" -f
  # ^ Requests the current page, saving it as "packages.json" under  folder named ${page_num}, for example: the folder could be named "5".
  # We will also save the HTTP response headers for that page of API response in the same folder, as a separate file called "headers".
  # curl's "-f" flag makes it exit with non-zero exit status if the HTTP response status code is an error status, such as 500.
  # So we can try again later if we get an error HTT response from the Atom package API server... Which happened to me a lot IRL.
  # NOTE: I chose to back up the pages as sorted by creation date, as that should be a monotonic, stable format, with new data appended
  # only to the last page/any future pages. Sort by stars (the default) shows which packages are "the most important historically, of all time",
  # which is indeed a valuable way to sort, but that sort order could be unstable,
  # making ongoing incremental backups of the data much more complicated and less tidy.

  curl_exit_status=${?} # Store the exit status of the curl command above so we can branch based on HTTP response code "error" or "success"
  let "attempts+=1" # Iterate this var so we can print a nice log of how many attempts we've made so far to fetch this page.

  sleep 15s # Let's be nice to the Atom package API... Only make another network request to the API every 15 seconds. Yes, this is slow. But it is only polite to the other Atom users.
 done

done
```

</details>

**Reply by confused-Techie** [reply]
@DeeDeeG fantastic bash script, smart bit about the waiting between each request, something I should probably implement. You are absolutely correct that a server error caused it to crash this morning, but that has now been resolved, to waiting 2 seconds multiplied by the amount of consecutive fails up to 50 fails, wherein it will exit. But this script absolutely does the job. I've already forked the repo full of documentation for Atom.io, which seems to truly contain everything, and after that we just want to collect the standard webpages themselves. 

Once this is said and done, the work could begin to recreate the API backend. Unless you had any ideas or insight into reverse engineering the backend, I was thinking to start from scratch and work towards feature parity as quickly as possible.

**Reply by DeeDeeG** [reply]
Yeah, I think re-implementing the API server logic from scratch would be the only way, if only because I don't think the server code was ever open-sourced, and there aren't devs responding to any of the issues over at https://github.com/atom/atom, I think the community are effectively on their own.

Edit to add: At least the [documentation](https://flight-manual.atom.io/atom-server-side-apis/sections/atom-package-server-api/) is clear as to how the API should respond. And the official server is still up, so if we have remaining questions about how the actual official implementation works, we can poke the official API a bit and find out.

(We could open a separate, low-traffic issue or discussion post and try to ping past Atom team members and try to get code, [and/or thoughts on intellectual property of the data???] for the official API server.)

**Reply by DeeDeeG** [reply]
Miscellaneous thoughts I'm having:

I do want to be up-front: I don't have much strong coding skills myself. I'm happy to provide my knowledge of having seen and troubleshooted lots of the code. But architecturing/authoring a lot of novel code isn't something I can do for this effort.

If you want to go for it, I'll try to be helpful in whatever way I can. (Anyone else sitting back and reading, thinking "I could do that!", feel free to jump in and offer what you have in the way of skills as well!)

So looking forward: there are some implications for re-implementing the API server.

Work/to-do items for Atom core and apm: (switch the hard-coded API URL from `atom.io` to whatever the new URL would be. I think I could manage that. And I would want to look at https://www.npmjs.com/package/atom-package-dependencies to make sure that still works. In any case, the "community fork" of the above (Atom, apm, a few packages) could be updated to use the new URL and it should be fine.

There's also the sort of "intellectual property" question of who owns the old package metadata, and the sense of an implicit contract package authors had with a trusted entity of GitHub. Is it cool for us to plug that metadata into the new community server? Will package authors get bothered by support requests with our somewhat non-standard (not by much, but still) forked Atom? Do we need to start from scratch and re-populate the community package API database from zero? Does GitHub mind us using the old data/packages list? Do package authors mind? Kind of unclear. I think having pure intentions is a good start (we do), communicate openly and welcome anyone's questions or concerns, be responsive, and wait to see if anyone has a problem with it? Hmm. (Other than starting "clean-room" from scratch, there's no other way around it but being open and taking feedback on it.)

**Reply by confused-Techie** [reply]
All great thoughts, and I definitely think a new Discussion might be best to focus on this. As for the code, I have already started laying some bare bones groundwork, available at my repo [Atom-Community-Server-Backend](https://github.com/confused-Techie/atom-community-server-backend). But reaching out for some information, especially how auth was handled, and the type of search algorithm that was implemented for `/api/package/search` to help ensure as consistent as possible thoughts.

As for the thought of who owns the package metadata that's a hard one. But there are some solid approaches you've outlined:
* We could reach out to all of the package owners and directly ask them if this is something they are open to. This of course would be a huge undertaking but is one method.
* While starting from scratch may be the safest approach to avoid any negative repercussions from package owners and/or GitHub one huge concern is that I think a large part of Atom's quality is derived from the availability of packages, granted (IIRC) about 50 Atom native packages are installed by default and these could very well be Open Source already, that would still be a huge loss to the ecosystem starting from scratch.
* If we did decide to charge ahead with the usage of said data, we could have a period or some defined method for removal of this data, to make it as easy as possible for any package publishers that are uncomfortable with this to ask for us to remove their data.
* And finally the inverse of the above suggestion. If we do move forward with the data, its on a time limit to be removed unless given express permission by Package Owners.

In the meantime luckily much of work can continue to lay the groundwork without knowing the answers here. But it is absolutely something that needs to be addressed.

Additionally, like you mentioned, if anyone is reading our thread thinking to themselves they can do this or want to help, please feel free to start your version or contribute to my existing one, anything helps.

**Reply by kaosine** [reply]
Might be useful to use this, this seems to partially document it. (who knew searching api off the atom org would help?) https://flight-manual.atom.io/atom-server-side-apis/

Whether or not that's of any use to anyone that can get it done before the deadline is another story.

**Reply by kaosine** [reply]
Also might be useful to look at this new reply where they're supposedly discussing this in a atom community server as well on discord https://github.com/atom-community/atom/discussions/361#discussioncomment-2958735

**Comment by kaosine** [comment]
I'd honestly like to see if this would continue here despite the sunset of the official atom repositories. I noticed the comment that basically linked me to this community repo and I'm certain it can't be hard with all the interest in the offical repos issues(that have blown up my email since the announcemnent) to find people that can maintain it on this side or at least I'd hope so. I wish I could offer myself up but I'm not sure I'd be the best person for that since I don't have nearly enough knowledge to program let alone lead something like this. It'd be nice to hear something from the people who are maintaining these forks and have been doing so from the looks of it and being a bit more active than MS's leftover team they had on atom....

**Comment by ghost** [comment]
I would like to support efforts on this project here and there as I can, but cannot at this time devote significant technical skill. I do have some ideas for reworking the CI pipeline and moving it over to github actions (or perhaps Gitlab). I hesitate, however, to spearhead such an effort - as I am still teaching myself many things about CI pipelines. I will check in here and there as I can... but am by no means a professional nor should I be regarded as a primary contributor to the project moving forward.

**Reply by kaosine** [reply]
Definitely should be using actions, however I'm still trying to teach myself those with pages so I can get some custom stuff on mine....which that could possibly be how the site is handled too but idk. Not like it'd need much other than a external source for the api database...which it looks like pretty much could just be a request to update a data file in any number of these static pages generators.

**Reply by aminya** [reply]
> reworking the CI pipeline and moving it over to github actions (or perhaps Gitlab

I don't think moving things to GitHub Actions helps the end-users. I redesigned the whole Azure CI system when we created this fork. I made everything reusable templates and unified the scripts to use Powershell on all platforms. Many of the changes were upstreamed. Moving to new technologies takes time, and such decisions do not make a noticeable difference for the end-user.

https://github.com/atom-community/atom/tree/master/script/vsts

https://github.com/atom-community/atom/tree/master/script/vsts/platforms/templates

**Reply by joshcangit** [reply]
To me, a major problem with GitHub Actions is that I have to be signed in to download the artifacts.
I think the CI should be publicly available.

**Comment by mauricioszabo** [comment]
Folks, we're discussing on Discord (https://discord.gg/CdSBwZMU) some of the challenges and what we can do. I'm currently working on making the whole process of "compiling a working version" of Atom more easily available (no outdated node, no strange packages, no failures while running `npm install`, no `./script/bootstrap` that have to be maintained and also sometimes do not work), so if anyone wants to join, let's talk there :)

Also, thanks for keeping the discussions open here, so we have a public place to know that we may not be dead yet :D

**Comment by Digitalone1** [comment]
I have a website project written in php and some of its functionality are provided by javascript code executed on the browser (I entirely developed it inside atom).

So I have some javascript skills, even if a big project like atom is a very different thing from what I do with my website.

Anyway I feel like I could contribute somehow, but there are some issues at the moment. There's no roadmap or a list of things to do. For example, imagine there are some functions to change for a new naming convention, which you don't need very specific skills, I can change them in 5 seconds with atom search and replace feature through project workspace, but I'm not aware of it.

There should be a place here to list those things. Since you have to use Github.com because it's tied to the codebase and migrating to Gitlab it's too complex, just use all the features of this platform.

I know you're on Discord, but honestly, it's not used by everyone and its app is one of the worst on Linux, I'd use it in the browser, but it's not really convenient in my opinion.

Another issue is that I don't have any clue on how to test this project. Even if I don't want to contribute, just for curiosity, I have Arch Linux but I can't build it because it's broken. There's an AUR PKGBUILD not working.

I'm currently using the old Arch upstream packages which were deleted after the sunset. 

I'm contributing to EasyEffects on Linux and my motivation is that it's quite easy just because it has a working PKGBUILD on AUR, even if I compile the testing build outside pacman.

Anyway, in the end, having a roadmap here with list of things to do would be a good starting point.

**Reply by kaosine** [reply]
> Anyway I feel like I could contribute somehow, but there are some issues at the moment. There's no roadmap or a list of things to do. For example, imagine there are some functions to change for a new naming convention, which you don't need very specific skills, I can change them in 5 seconds with atom search and replace feature through project workspace, but I'm not aware of it.

Well that's one thing I've been trying to ask about fixing. I know right now @mauricioszabo has been working on some updates to make the editors core a bit more up to date. I'd like to see the projects section converted to the more modern interface(it's using the old version and is a bit out of date since it fell out of use for awhile). I'd like to see a proper roadmap too but I think at the moment we're stuck at getting it from electron v11(which Mauricio got it to 12) to as close to up to date as possible. 


> I know you're on Discord, but honestly, it's not used by everyone and its app is one of the worst on Linux, I'd use it in the browser, but it's not really convenient in my opinion.

Well doesn't help that the only other decent alternative which is matrix imo is lacking a bit in it's organization features. With what changes probably need to be made in the discord for ease of use it's easier to use that and the discussion tab here afaik. But that's a question for the ones with read/write access(namely @aminya @DeeDeeG and ice cream(idk what their name is here)

> Another issue is that I don't have any clue on how to test this project. Even if I don't want to contribute, just for curiosity, I have Arch Linux but I can't build it because it's broken. There's an AUR PKGBUILD not working.

This too is probably fixed in the pr that @mauricioszabo has open as well. I know he supposedly got builds working again-ish with that. 

A definite roadmap tho wouldn't be a bad idea but I know we're still debating whether a rename is in order(it might but not entirely sure). It'd help a lot to build some kind of team that has a couple people that are responsible for each major part(api, core, frontend, docs, etc). Which I've got a personal repo that will be pushed towards the website/docs side of things hopefully soon. I know the state things are in now we're in the 1.63-dev side of things from upstream but with the amount of work to be done bumping to 2.0 might be nice to give a clear end to atoms side and where we begin with going forward.

**Reply by mauricioszabo** [reply]
There are low-hanging fruits here on the Atom API that needs to be done: https://github.com/confused-Techie/atom-community-server-backend-JS/issues

One of the **really bad** parts of Atom is that it's not easy to build from source. That's actually what I'm doing right now: trying to make build from source works better: https://github.com/AtomRebirth/atom/tree/build-without-scripts. I tried to build some packages with CircleCI, but I quickly got into its monthly limits so I'm trying Cirrus-CI now, to at least be able to produce binaries easily

**Reply by confused-Techie** [reply]
Absolutely feel free to work on some of the issues on the Atom-Community-Server-Backend-JS, I tried to make it as easy as possible to understand to anyone that wanted to help. Plus feel free to look through the README as I've done what I can to make it as easy as possible for others to contribute. Like a list of all [functions/methods/routes](https://github.com/confused-Techie/atom-community-server-backend-JS/blob/main/docs/major_todo.md) that are not yet finished, and need work. Or if more advanced [complexity reports](https://github.com/confused-Techie/atom-community-server-backend-JS/blob/main/docs/complexity-report.md) to help know where may need more thorough testing. But even pointing out issues that still need to be worked on is helpful, to help get someone to look that direction. Anything at this point is appreciated
